javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/collector/ReferenceCollector.java

```diff
@@ -7,54 +7,103 @@
 
 import static com.google.common.base.Preconditions.checkNotNull;
 import static io.opentelemetry.javaagent.tooling.muzzle.InstrumentationClassPredicate.isInstrumentationClass;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Collections.singleton;
 
+import com.google.common.base.Strings;
 import com.google.common.graph.Graph;
 import com.google.common.graph.GraphBuilder;
 import com.google.common.graph.Graphs;
 import com.google.common.graph.MutableGraph;
 import io.opentelemetry.javaagent.tooling.Utils;
 import io.opentelemetry.javaagent.tooling.muzzle.Reference;
+import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.InputStreamReader;
 import java.net.URLConnection;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Queue;
 import java.util.Set;
+import java.util.regex.Pattern;
 import net.bytebuddy.jar.asm.ClassReader;
 
 /**
  * {@link LinkedHashMap} is used for reference map to guarantee a deterministic order of iteration,
  * so that bytecode generated based on it would also be deterministic.
+ *
+ * <p>This class is only called at compile time by the {@link MuzzleCodeGenerationPlugin} ByteBuddy
+ * plugin.
  */
 public class ReferenceCollector {
   private final Map<String, Reference> references = new LinkedHashMap<>();
   private final MutableGraph<String> helperSuperClassGraph = GraphBuilder.directed().build();
   private final Set<String> visitedClasses = new HashSet<>();
 
+  /**
+   * If passed {@code resource} path points to an SPI file (either Java {@link
+   * java.util.ServiceLoader} or AWS SDK {@code ExecutionInterceptor}) reads the file and adds every
+   * implementation as a reference, traversing the graph of classes until a non-instrumentation
+   * (external) class is encountered.
+   *
+   * @param resource path to the resource file, same as in {@link ClassLoader#getResource(String)}
+   * @see io.opentelemetry.javaagent.tooling.muzzle.InstrumentationClassPredicate
+   */
+  public void collectReferencesFromResource(String resource) {
+    if (!isSpiFile(resource)) {
+      return;
+    }
+
+    List<String> spiImplementations = new ArrayList<>();
+    try (InputStream stream = getResourceStream(resource)) {
```

Comment:
Is it possible to use `Files.lines` or `Files.readAllLines`?

Reply:
I wanted to use one of those but they all accept `Path` - and we pretty much have to use `InputStream` here because of `connection.setUseCaches(false)`